### PR TITLE
[DGL-KE] Clean up argument for DGL-KE

### PR DESCRIPTION
**Pull request by aksnzhy**
## Description

Clean up arguments for DGL-KE.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: apps/kg/eval.py [hunk 1]
```diff
@@ -105,102 +94,74 @@ def main(args):
...
+                                                        args.neg_sample_size_test,
+                                                        args.neg_sample_size_test,
                                                         args.eval_filter,
                                                         mode='chunk-head',
-                                                        num_workers=num_workers,
-                                                        rank=0, ranks=1)
-        test_sampler_tail = eval_dataset.create_sampler('test', args.batch_size,
-                                                        args.neg_sample_size,
-                                                        args.neg_chunk_size,
+                                                        num_workers=1,
```
Comment (zheng-da) [inline apps/kg/eval.py]: why `num_workers` is 1 here?

File: apps/kg/eval.py [hunk 2]
```diff
@@ -105,102 +94,74 @@ def main(args):
...
-        for metric in logs[0].keys():
-            metrics[metric] = sum([log[metric] for log in logs]) / len(logs)
-        for k, v in metrics.items():
-            print('Test average {} at [{}/{}]: {}'.format(k, args.step, args.max_step, v))
-
-        for proc in procs:
-            proc.join()
-    else:
-        test(args, model, [test_sampler_head, test_sampler_tail])
+    #if args.num_proc > 1:
```
Comment (zheng-da) [inline apps/kg/eval.py]: let's not always create a new process to run eval. It's more difficult to debug when running the computation in a different process.

File: apps/kg/kvclient.py [hunk 3]
```diff
@@ -35,46 +32,34 @@ def __init__(self):
                           help='dataset name, under data_path')
         self.add_argument('--format', type=str, default='1',
                           help='the format of the dataset.')
-        self.add_argument('--save_path', type=str, default='ckpts',
+        self.add_argument('--save_path', type=str, default='../ckpts',
```
Comment (zheng-da) [inline apps/kg/kvclient.py]: why the default path is in the parent directory?

File: apps/kg/kvclient.py [hunk 4]
```diff
@@ -191,7 +166,8 @@ def get_local_machine_id(server_namebook):
 def start_worker(args, logger):
     """Start kvclient for training
     """
-    train_time_start = time.time()
+    init_time_start = time.time()
+    time.sleep(5) # wait for launch script
```
Comment (zheng-da) [inline apps/kg/kvclient.py]: is it always successful?

File: apps/kg/kvclient.py [hunk 5]
```diff
@@ -218,50 +194,43 @@ def start_worker(args, logger):
...
 
     train_samplers = []
     for i in range(args.num_client):
         train_sampler_head = train_data.create_sampler(args.batch_size,
                                                        args.neg_sample_size,
-                                                       args.neg_chunk_size,
+                                                       args.neg_sample_size,
                                                        mode='head',
-                                                       num_workers=num_workers,
+                                                       num_workers=1,
```
Comment (zheng-da) [inline apps/kg/kvclient.py]: is `num_workers=1` always the best option?

File: apps/kg/train.py [hunk 6]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-    else:
+    train_samplers = []
+    for i in range(args.num_proc):
         train_sampler_head = train_data.create_sampler(args.batch_size,
                                                        args.neg_sample_size,
-                                                       args.neg_chunk_size,
+                                                       args.neg_sample_size,
                                                        mode='head',
-                                                       num_workers=num_workers,
+                                                       num_workers=4,
```
Comment (zheng-da) [inline apps/kg/train.py]: maybe a larger number. for example, 8.

File: apps/kg/train.py [hunk 7]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                       exclude_positive=False)
+                                                       exclude_positive=False,
+                                                       rank=i)
         train_sampler_tail = train_data.create_sampler(args.batch_size,
                                                        args.neg_sample_size,
-                                                       args.neg_chunk_size,
+                                                       args.neg_sample_size,
                                                        mode='tail',
-                                                       num_workers=num_workers,
+                                                       num_workers=4,
```
Comment (zheng-da) [inline apps/kg/train.py]: the same here.

File: apps/kg/train.py [hunk 8]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                             args.neg_chunk_size_valid,
-                                                             args.eval_filter,
-                                                             mode='chunk-head',
-                                                             num_workers=num_workers,
-                                                             rank=0, ranks=1)
+                                                              args.neg_sample_size_valid,
+                                                              args.neg_sample_size_valid,
+                                                              args.eval_filter,
+                                                              mode='chunk-head',
+                                                              num_workers=1,
```
Comment (zheng-da) [inline apps/kg/train.py]: why is it 1 here?

File: apps/kg/train.py [hunk 9]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                             args.neg_chunk_size_valid,
-                                                             args.eval_filter,
-                                                             mode='chunk-tail',
-                                                             num_workers=num_workers,
-                                                             rank=0, ranks=1)
+                                                              args.neg_sample_size_valid,
+                                                              args.neg_sample_size_valid,
+                                                              args.eval_filter,
+                                                              mode='chunk-tail',
+                                                              num_workers=1,
```
Comment (zheng-da) [inline apps/kg/train.py]: the same here.

File: apps/kg/train.py [hunk 10]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                            args.neg_chunk_size_test,
-                                                            args.eval_filter,
-                                                            mode='chunk-head',
-                                                            num_workers=num_workers,
-                                                            rank=0, ranks=1)
+                                                             args.neg_sample_size_test,
+                                                             args.neg_sample_size_test,
+                                                             args.eval_filter,
+                                                             mode='chunk-head',
+                                                             num_workers=1,
```
Comment (zheng-da) [inline apps/kg/train.py]: the same here.

File: apps/kg/train.py [hunk 11]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                            args.neg_chunk_size_test,
-                                                            args.eval_filter,
-                                                            mode='chunk-tail',
-                                                            num_workers=num_workers,
-                                                            rank=0, ranks=1)
+                                                             args.neg_sample_size_test,
+                                                             args.neg_sample_size_test,
+                                                             args.eval_filter,
+                                                             mode='chunk-tail',
+                                                             num_workers=1,
```
Comment (zheng-da) [inline apps/kg/train.py]: the same here.

File: apps/kg/train.py [hunk 12]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                     barrier))
-            procs.append(proc)
-            proc.start()
-        for proc in procs:
-            proc.join()
-    else:
-        valid_samplers = [valid_sampler_head, valid_sampler_tail] if args.valid else None
-        train(args, model, train_sampler, valid_samplers, rel_parts=rel_parts)
+    procs = []
+    barrier = mp.Barrier(args.num_proc)
```
Comment (zheng-da) [inline apps/kg/train.py]: it's better to not always run training in other processes.

File: apps/kg/train_pytorch.py [hunk 13]
```diff
@@ -279,10 +275,8 @@ def dist_train_test(args, model, train_sampler, entity_pb, relation_pb, l2g, ran
             end += count
             percent += 1
 
-        if args.save_emb is not None:
-            if not os.path.exists(args.save_emb):
-                os.mkdir(args.save_emb)
```
Comment (zheng-da) [inline apps/kg/train_pytorch.py]: why remove this?

**Comment by zheng-da** [comment]
i'm still confused what is the interface of distributed training?

File: apps/kg/eval.py (hunk 2 above)
Comment (classicsong) [reply apps/kg/eval.py]: Agree, if num_proc == 1, we can avoid fork.

File: apps/kg/eval.py [hunk 14]
```diff
@@ -105,102 +94,74 @@ def main(args):
...
-            test_sampler_heads.append(test_sampler_head)
-            test_sampler_tails.append(test_sampler_tail)
-    else:
-        test_sampler_head = eval_dataset.create_sampler('test', args.batch_size,
-                                                        args.neg_sample_size,
-                                                        args.neg_chunk_size,
+
+    test_sampler_tails = []
+    test_sampler_heads = []
+    for i in range(args.num_proc):
```
Comment (classicsong) [inline apps/kg/eval.py]: we should assert args.num_proc >= 1

File: apps/kg/train.py [hunk 15]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-                                                           num_workers=num_workers,
-                                                           shuffle=True,
-                                                           exclude_positive=False,
-                                                           rank=i)
-            train_samplers.append(NewBidirectionalOneShotIterator(train_sampler_head, train_sampler_tail,
-                                                                  args.neg_chunk_size, args.neg_sample_size,
-                                                                  True, n_entities))
-    else:
+    train_samplers = []
+    for i in range(args.num_proc):
```
Comment (classicsong) [inline apps/kg/train.py]: we should check args.num_proc >= 1

File: apps/kg/train.py [hunk 16]
```diff
@@ -153,250 +129,160 @@ def get_logger(args):
...
-    eval_dataset = None
-    dataset = None
     # load model
-    model = load_model(logger, args, n_entities, n_relations)
+    model = load_model(logger, args, dataset.n_entities, dataset.n_relations)
+    model.share_memory()
 
-    if args.num_proc > 1 or args.async_update:
-        model.share_memory()
+    eval_dataset = None
```
Comment (classicsong) [inline apps/kg/train.py]: Add a comment here. This is used to free memory

File: apps/kg/eval.py (hunk 1 above)
Comment (aksnzhy) [reply apps/kg/eval.py]: I will fix all the num_workers to 8.

File: apps/kg/kvclient.py (hunk 3 above)
Comment (aksnzhy) [reply apps/kg/kvclient.py]: Because the distributed task runs on sub-directory.

File: apps/kg/kvclient.py (hunk 4 above)
Comment (aksnzhy) [reply apps/kg/kvclient.py]: I change it to WAIT_TIME, which is defined as 10 seconds. I tested it for ssh 200 nodes and it is no problem.

File: apps/kg/train_pytorch.py (hunk 13 above)
Comment (aksnzhy) [reply apps/kg/train_pytorch.py]: It is not been removed. I change it to :

```
if args.save_emb is True:
            model_test.save_emb(args.save_path, args.dataset)
```

**Comment by aksnzhy** [comment]
> i'm still confused what is the interface of distributed training?

1. Change the `ip_config.txt`
2. Run `distributed/launch.sh`

Use can modify the parameters in `launch.sh`

```
script_path=~/dgl/apps/kg/distributed
script_file=./freebase_transe_l2.sh
user_name=ubuntu
ssh_key=~/mctt.pem
```

File: apps/kg/train_pytorch.py (hunk 13 above)
Comment (zheng-da) [reply apps/kg/train_pytorch.py]: what i mean is why removing the code of creating a directory? it's possible that the directory doesn't exist.

**Comment by zheng-da** [comment]
i guess it's better that `launch.sh` accepts arguments.

File: apps/kg/train_pytorch.py (hunk 13 above)
Comment (aksnzhy) [reply apps/kg/train_pytorch.py]: I find we must use save-emb url in test. Just change it back...

**Comment by aksnzhy** [comment]
Sure. I will change it in another PR.

**Pull request merged**